ec/src/msm/variable_base/stream_pippenger.rs

```diff
@@ -65,3 +65,66 @@ impl<G: AffineCurve> ChunkedPippenger<G> {
         self.result
     }
 }
+
+/// Hash map struct for Pippenger algorithm.
+pub struct HashMapPippenger<G: AffineCurve> {
+    pub buffer: HashMap<G, G::ScalarField>,
+    pub result: G::Projective,
+}
+
+impl<G: AffineCurve> HashMapPippenger<G> {
+    /// Produce a new hash map with the maximum msm buffer size.
+    pub fn new(max_msm_buffer: usize) -> Self {
+        Self {
+            buffer: HashMap::with_capacity(max_msm_buffer),
+            result: G::Projective::zero(),
+        }
+    }
+
+    /// Add a new (base, scalar) pair into the hash map.
+    #[inline(always)]
+    pub fn add<B, S>(&mut self, base: B, scalar: S)
+    where
+        B: Borrow<G>,
+        S: Borrow<G::ScalarField>,
+    {
+        // update the entry, guarding the possibility that it has been already set.
+        let entry = self
+            .buffer
+            .entry(*base.borrow())
+            .or_insert(G::ScalarField::zero());
+        *entry += *scalar.borrow();
+        if self.buffer.len() == self.buffer.capacity() {
```

Comment:
@huyuncong @mmaker I have a concern about whether `capacity()` will run in the expected way. Rust did not promise that the capacity of the map would not grow on its own, and Rust did not promise that it will allocate exactly the capacity as specified during `with_capacity`.

Note: Rust's hashmap is often constructed through the RawTable here, which may specify a slightly different capacity compared with the one from `with_capacity`.
https://docs.rs/hashbrown/latest/src/hashbrown/raw/mod.rs.html#191